File: packages/evals/tasks/bench/observe/panamcs.ts
import { defineBenchTask } from "../../../framework/defineTask.js";
import { findMatchingSelector } from "../../../framework/observeSelectors.js";

export default defineBenchTask(
  { name: "panamcs" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/panamcs/");

      const { data: observations } = await stagehand.observe("click the 'about us' link");

      if (observations.length === 0) {
        return {
          _success: false,
          observations,
          debugUrl,
          sessionUrl,
          logs: logger.getLogs(),
        };
      }

      let foundMatch = false;
      for (const observation of observations) {
        try {
          if (
            await findMatchingSelector(page, observation.selector, ["#menu > li:nth-child(1) > a"])
          ) {
            foundMatch = true;
            break;
          }
        } catch (error) {
          console.warn(
            `Failed to check observation with selector ${observation.selector}:`,
            error instanceof Error ? error.message : String(error),
          );
          continue;
        }
      }

      return {
        _success: foundMatch,
        expected: "#menu > li:nth-child(1) > a",
        observations,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/observe/panamcs.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/observe/panamcs.ts

Evals.define_task("panamcs") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/panamcs/")

  observations = t.stagehand.observe("click the 'about us' link").data

  if observations.empty?
    next { _success: false, observations: observations.map(&:to_wire) }
  end

  expected_locator = "#menu > li:nth-child(1) > a"

  expected_result = t.page.locator(expected_locator).first.inner_text

  found_match = observations.any? do |observation|
    t.page.locator(observation.selector).first.inner_text == expected_result
  rescue StandardError => e
    t.logger.log("Failed to check observation with selector #{observation.selector}: #{e.message}")
    false
  end

  { _success: found_match, expected: expected_result, observations: observations.map(&:to_wire) }
end
